gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -252,24 +255,52 @@ private WriteObjectResponse doResumableUpload() throws IOException {
 
       boolean objectFinalized = false;
       while (!objectFinalized) {
-        WriteObjectRequest insertRequest;
+        WriteObjectRequest insertRequest = null;
         if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
           insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
           writeOffset += insertRequest.getChecksummedData().getContent().size();
         } else {
-          ByteString data =
-              ByteString.readFrom(
-                  ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE), MAX_BYTES_PER_MESSAGE);
-          dataChunkMap.put(writeOffset, data);
           if (dataChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
-            dataChunkMap.remove(dataChunkMap.firstKey());
+            /*
+             If dataChunkMap is full, get the committedWriteOffset. This will make a API call to the
+             server and add latency in this path/context. Since there are already chunks in flight,
+             calling this API will not reduce overall throughput. It will throttle the upstream
+             write call now rather than onFinalize, which is fine. This will also increase QPS to
+             the GCS backend. The increase will be linear to number of chunks written, so that
+             should also be fine.
+            */
+            committedWriteOffset = getCommittedWriteSizeWithRetries(uploadId);
+            logger.atFinest().log(
+                "Fetched committedWriteOffset: size:%d, numBuffers:%d, writeOffset:%d, committedWriteOffset:%d",
+                dataChunkMap.size(),
+                channelOptions.getNumberOfBufferedRequests(),
+                this.writeOffset,
+                committedWriteOffset);
+
+            // check and remove chunks from dataChunkMap
+            while (dataChunkMap.size() > 0 && dataChunkMap.firstKey() < committedWriteOffset) {
+              logger.atFinest().log(
+                  "clearing dataChunkMap one buffer at a time, size: %d, firstKey:%d, committedwriteOffset:%d",
+                  dataChunkMap.size(), dataChunkMap.firstKey(), committedWriteOffset);
+              dataChunkMap.remove(dataChunkMap.firstKey());
```

Comment:
Is it safe to remove data from the map while iterating over it ? may be it is safer to use iterator/ stream API

```
map.entrySet().removeIf(entry -> entry.getKey() < committedWriteOffset));
```

OR
```
Iterator<Map.Entry<String, Integer>> it = map.entrySet().iterator();
        while (it.hasNext()) {
            if (it.next().getKey() < committedWriteOffset)){
                it.remove();
            } else {
               break;
           }
        }
```

Reply:
Current implementation is safe. We are not really iterating, we are looking at the head(`firstKey`) and removing it if it is committed.